src/tools/illink/src/ILLink.Shared/TrimAnalysis/ValueExtensions.cs

```diff
@@ -50,10 +52,20 @@ internal static string ValueToString (this SingleValue value, params object[] ar
 
 		internal static SingleValue? AsSingleValue (this in MultiValue node)
 		{
-			if (node.Count () != 1)
+			var values = node.AsEnumerable ();
+			if (values.Count () != 1)
 				return null;
 
-			return node.Single ();
+			return values.Single ();
+		}
+
+		// ValueSet<TValue> is not enumerable. This helper translates ValueSet<SingleValue>.Unknown
+		// into a ValueSet<SingleValue> whose sole element is UnknownValue.Instance.
+		internal static IEnumerable<SingleValue> AsEnumerable (this MultiValue multiValue)
```

Comment:
I must admit I'm not super excited about the need for this... but I also don't know about a better way - other than teaching ValueSet about the "error" value for all value sets (which is not ideal either).

Reply:
The other disadvantage of this approach is that each enumeration will now allocate, actually twice, once to box the enumerable and then to create the enumerator. If value set implement enumerable, it can rely on the compiler optimization to avoid these instead.

Reply:
One other approach might be this (not pretty, but might work):
Make MultiValue a wrapper (since struct can't inherit) - so declare MultiValue as a real type, and make it wrap the ValueSet. And them make it implement IEnumerable directly.
Not sure if there are places though where we heavily rely on the ability to operate on ValueSets directly. On the other hand the conversion could be implicit (since wrapping doesn't induce allocation), so it might be possible handle this nicely almost everywhere.

Reply:
Or (just idea):
* Change ValueSetLattice to a class (do we need this to be a struct, does it make much of a difference?)
  * There some weird things about the lattice as is now
  * It's a struct, but we use it through an interface -> boxing it anyway
  * It's a struct, so can't be derived, thus the implementation of all of its methods is effectively `static`.
  * For other lattices it's benefitial to be instance based (we need it in some cases), but for the ValueSetLattice this seems unnecessary, so having it provide the Top/Bottom as statics should work (alongside the interface implementations). Or maybe it would just provide the singleton value (it could be a singleton, we don't need to allocate more than one since they're all the same).
* Have ValueSet be a generic over TLattice
* Make the "Error" value a static on the lattice - or more precisely, make the ValueSetLattice define both Top and Bottom (only the ValueSetLattice for now, we would not need it on all lattices).

This would make the design overall slightly cleaner. For example now we rely on the implicit rule that the `default` value of a ValueSet is the `Top` value of the lattice.